File: ros2_workspaces/ros2_from_scratch_by_Renard_ws/src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from my_robot_interfaces.srv import ResetCounter


"""
NumberCounterNode increments `counter` value by the values consumed from the `number topic`. 
Thus, NumberCounterNode is a subscriber to `number`. 
NumberCounterNode exposes `reset_counter` service to reset the counter value to a value between 0 and the current value.
"""
class NumberCounterNode(Node): 
    def __init__(self):
        super().__init__("number_counter") 
        self.get_logger().info("Starting " + self.get_name() + "...")
        self.counter_ = 0
        # author recommends best practice of nameing callbacks `callback_<topic>`
        self.number_subscriber_ = self.create_subscription(Int64, "number", self.callback_number, 10)
                                                          # interface,  service name     service callback
        self.reset_counter_service_ = self.create_service(ResetCounter, "reset_counter", self.callback_reset_counter)
        self.get_logger().info(self.get_name() + " started!!")

    def callback_reset_counter(self, request:ResetCounter.Request, response: ResetCounter.Response):
        if request.reset_value < 0:
            response.success = False
            response.message = "Cannot reset counter to a negative value"
        elif request.reset_value > self.counter_:
            response.success = False
            response.message = "Reset value must be lower than current counter value"
        else:
            self.counter_ = request.reset_value
            self.get_logger().info("Reset counter to " + str(self.counter_))
            response.success = True
            response.message = "Success"
        return response

    def callback_number(self, msg:Int64):
        self.counter_ += msg.data
        self.get_logger().info("Counter: " + str(self.counter_))

# entry point
def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_workspaces/Books/ros2_from_scratch_by_Renard/part2_ws/src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "my_robot_interfaces/srv/reset_counter.hpp"

using namespace std::placeholders;
using ResetCounter = my_robot_interfaces::srv::ResetCounter;
using Int64 = example_interfaces::msg::Int64;

class NumberCounterNode : public rclcpp::Node {

    public:
    /** 
     * variable initialization for parent and child classes in the `member initializer list`.
     * Initializes members at the exact moment they are created. 
     * Mandatory for base class constructors.
    */
    NumberCounterNode() : Node("number_counter_node_cpp"), counter_(0) {
        /**
         * May initialize inside because we require callbackNumber.
         */
        number_subscriber_ = this->create_subscription<Int64> (
            "number", 10, std::bind(&NumberCounterNode::callbackNumber, this, _1));
        /**
         * 
         */
        reset_counter_service_ = this->create_service<ResetCounter>("reset_counter", std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
        
        RCLCPP_INFO(this->get_logger(), "Number counter is created!!");
    }

    private:
    // service callback
    /* Note the shared pointers instead of a reference */
    void callbackResetCounter(const ResetCounter::Request::SharedPtr request, const::ResetCounter::Response::SharedPtr response) {
        if (request->reset_value < 0) {
            response->success = false;
            response->message = "Cannot resent counter to a negative value";
        } else if (request->reset_value > counter_) {
            response->success = false;
            response->message = "Reset value must be lower than the current counter value";
        } else {
            counter_ = request->reset_value;
            RCLCPP_INFO(this->get_logger(), "Reset counter to %d", counter_);
            response->success = true;
            response->message = "Success";
        }
    }
    // increment counter by the message value, log it.
    void callbackNumber(const Int64::SharedPtr msg) {
        counter_ += msg->data;
        RCLCPP_INFO(this->get_logger(), "Counter: %d", counter_);
    }
    int counter_;
    rclcpp::Subscription<Int64>::SharedPtr number_subscriber_; // subscriber
    rclcpp::Service<ResetCounter>::SharedPtr reset_counter_service_; // service
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}